### PR TITLE
Add support for managed identities

### LINKED ISSUE #5: Can this be modified to support accessing blobs with MSI?
In our environment, we do not want to generate "root level" SAS tokens.

Is there anyway this plugin can be modified to support using a "bearer" token obtained from the OAuth endpoint?

### PR DESCRIPTION
**Pull request by elsesiy**
Fix #5

### CODE REVIEW

**Comment by elsesiy** [comment]
@microsoftopensource Please review

**Comment by elsesiy** [comment]
@logachev Can you please take a look here?

**Comment by elsesiy** [comment]
@microsoftopensource @logachev Any feedback?

**Comment by elsesiy** [comment]
@microsoftopensource @logachev It would be great to get this merged soon, thanks.

**Comment by logachev** [comment]
@elsesiy I'm sorry for the delay, was busy with another project.
I will take a look tomorrow! Thanks for doing this!

**Comment by elsesiy** [comment]
@logachev Great, thanks! Let me know if it needs work and I get it done right away

File: lib/fluent/plugin/out_azure-storage-append-blob.rb [hunk 1]
```diff
@@ -17,68 +18,112 @@ class AzureStorageAppendBlobOut < Fluent::Plugin::Output
...
-  
-        ensure_container
+        if @azure_use_managed_identity
+          token_credential = Azure::Storage::Common::Core::TokenCredential.new get_access_token
+          token_signer = Azure::Storage::Common::Core::Auth::TokenSigner.new token_credential
+          @bs = Azure::Storage::Blob::BlobService.new(storage_account_name: @azure_storage_account, signer: token_signer)
 
-        @azure_storage_path = ''
-        @last_azure_storage_path = ''
+          # Refresh internal is 50 minutes
```
Comment (logachev) [inline lib/fluent/plugin/out_azure-storage-append-blob.rb:98]: nit: typo, interval

File: lib/fluent/plugin/out_azure-storage-append-blob.rb [hunk 2]
```diff
@@ -17,68 +18,112 @@ class AzureStorageAppendBlobOut < Fluent::Plugin::Output
...
+          refresh_interval = 50 * 60
+          # The user-defined thread that renews the access token
+          cancelled = false
+          renew_token = Thread.new do
+            Thread.stop
```
Comment (logachev) [inline lib/fluent/plugin/out_azure-storage-append-blob.rb:103]: Sorry for my lack of Ruby experience.. What is this `Thread.stop` for? 
Isn't it immediate thread exit?

File: lib/fluent/plugin/out_azure-storage-append-blob.rb [hunk 3]
```diff
@@ -17,68 +18,112 @@ class AzureStorageAppendBlobOut < Fluent::Plugin::Output
...
+            while !cancelled
+              sleep(refresh_interval)
+              # Update the access token to the credential
+              token_credential.renew_token get_access_token
+            end
+          end
+          sleep 0.1 while renew_token.status != "sleep"
```
Comment (logachev) [inline lib/fluent/plugin/out_azure-storage-append-blob.rb:110]: Can you please also explain what this is doing?

File: lib/fluent/plugin/out_azure-storage-append-blob.rb (hunk 2 above)
Comment (elsesiy) [reply lib/fluent/plugin/out_azure-storage-append-blob.rb:103]: I'm not a ruby person either, this is taken straight from the official [azure-storage-ruby sdk docs](https://github.com/Azure/azure-storage-ruby/tree/master/blob#access-token)

File: lib/fluent/plugin/out_azure-storage-append-blob.rb (hunk 3 above)
Comment (elsesiy) [reply lib/fluent/plugin/out_azure-storage-append-blob.rb:110]: See previous answer, I can't really speak to this

File: lib/fluent/plugin/out_azure-storage-append-blob.rb (hunk 1 above)
Comment (elsesiy) [reply lib/fluent/plugin/out_azure-storage-append-blob.rb]: fixed

File: lib/fluent/plugin/out_azure-storage-append-blob.rb (hunk 2 above)
Comment (elsesiy) [reply lib/fluent/plugin/out_azure-storage-append-blob.rb:103]: @logachev I debugged this threading code and it works as expected. According to the documentation for [Thread.stop](https://ruby-doc.org/core-2.5.0/Thread.html) it puts the thread to sleep and resumes in the next line which enters the loop. I suspect it's done to unblock the rest of the code right after kicking off the thread.

File: lib/fluent/plugin/out_azure-storage-append-blob.rb (hunk 3 above)
Comment (elsesiy) [reply lib/fluent/plugin/out_azure-storage-append-blob.rb:110]: `Thread.stop` puts the thread to sleep so resumes the execution after

File: lib/fluent/plugin/out_azure-storage-append-blob.rb (hunk 2 above)
Comment (logachev) [reply lib/fluent/plugin/out_azure-storage-append-blob.rb:103]: Sounds good. Yeah, I was wondering why you did it this way, but it this is aligned with the SDK -- no further questions

**Comment by logachev** [comment]
@elsesiy is there a reason you put it to the draft?

**Comment by elsesiy** [comment]
@logachev Yes, please look at the referenced issue. This code is working as expected but there seems to be an issue in the azure-storage-ruby SDK where a token refresh doesn’t lead to a successful connection after the initial token expires

**Comment by elsesiy** [comment]
@logachev Please have another look at the PR

**Comment by elsesiy** [comment]
@logachev This PR is ready for review, can you have another look please?

File: lib/fluent/plugin/out_azure-storage-append-blob.rb [hunk 4]
```diff
@@ -3,122 +3,157 @@
...
       end
-  
+
       def multi_workers_ready?
         true
       end
-  
+
+      def get_access_token
+        access_key_request = Faraday.new('http://169.254.169.254/metadata/identity/oauth2/token?' \
```
Comment (wesyao) [inline lib/fluent/plugin/out_azure-storage-append-blob.rb:75]: this looks good for msi support, but i think the api version should be configurable

File: lib/fluent/plugin/out_azure-storage-append-blob.rb [hunk 5]
```diff
@@ -3,122 +3,157 @@
...
         super
-  
-        @bs_params = {storage_account_name: @azure_storage_account}
+        if @use_msi
+          token_credential = Azure::Storage::Common::Core::TokenCredential.new get_access_token
+          token_signer = Azure::Storage::Common::Core::Auth::TokenSigner.new token_credential
+          @bs = Azure::Storage::Blob::BlobService.new(storage_account_name: @azure_storage_account, signer: token_signer)
+
+          # Refresh interval is 50 minutes
+          refresh_interval = 50 * 60
```
Comment (wesyao) [inline lib/fluent/plugin/out_azure-storage-append-blob.rb:92]: refresh interval should be configurable but probably defaulted 60

File: lib/fluent/plugin/out_azure-storage-append-blob.rb (hunk 4 above)
Comment (elsesiy) [reply lib/fluent/plugin/out_azure-storage-append-blob.rb:75]: Sure, I'll add that.

File: lib/fluent/plugin/out_azure-storage-append-blob.rb (hunk 5 above)
Comment (elsesiy) [reply lib/fluent/plugin/out_azure-storage-append-blob.rb:92]: This is taken straight from the example provided in the azure storage ruby sdk, see [here](https://github.com/Azure/azure-storage-ruby/tree/master/blob#access-token). I can make it configurable but the tokens I have encountered so far are all valid for 23h59.

**Comment by elsesiy** [comment]
@wesyao I addressed your comments in the latest commit 67ff869, please let me know if you have any further remarks.

**Comment by wesyao** [comment]
lgtm

**Pull request merged**